### PR TITLE
App modules: add additional app properties such as app path, immersive process, and app architecture

### LINKED ISSUE #7894: App modules: add additional properties for benefit of testing, add-ons and alike
Hi,

Sprang from a discussion on Gitter:

A new add-on developer asks if a property to return the full path for an app be created. @JCSTeh gave a solution. And since I've been meaning to add a few more properties (namely detecting if we're dealing with a universal app or not), I though it would be best to raise a meta-issue here.

Thus please let us know if you need properties about an app that should be shown as part of an app module.

Thanks.

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #7894

### Summary of the issue:
Adds additional app properties such as path, immersive process, and architecture (WoW64).

### Description of how this pull request fixes the issue:
Adds the following properties:

* appPath: returns full executable path for the current application.
* isWindowsStoreApp: looks for a package info structure to determine if an app is a Store application (see a later comment as to why this route is used).
* appArchitecture: uses kernel32.dll's IsWoW64Process2 (Windows 10 Version 1511 and later) to return a string denoting which machien the app was written for, suitable for working with x86/x64 processes on ARM64.

### Testing performed:
Tested with various apps, both Win32 desktop apps and universal apps.

### Known issues with pull request:
None

### Change log entry:
Changes for developers:
Additional properties were added to app modules, including path for the executable (appPath), is a Windows Store app (isWindowsStoreApp), and machine architecture for the app (appArchitecture). (#7894)

### CODE REVIEW

**Comment by josephsl** [comment]
Hi,

I'll ask for a review once appArchitecture property is implemented and tested.

Thanks.

File: source/appModuleHandler.py [hunk 1]
```diff
@@ -444,6 +454,52 @@ def _get_is64BitProcess(self):
 		self.is64BitProcess = not res
 		return self.is64BitProcess
 
+	def _get_isWindowsStoreApp(self):
+		"""Whether this process is a Windows Store (immersive) process.
+		An immersive process is a Windows app that runs inside a Windows Runtime (WinRT) container.
+		These include Windows store apps on Windows 8 and 8.1, and Universal Windows Platform (UWP) apps on Windows 10.
+		@rtype: bool
+		"""
+		if winVersion.winVersion.major and winVersion.winVersion.minor == (6, 1):
```
Comment (leonardder) [inline source/appModuleHandler.py]: This line is wrong. It should be:

`if (winVersion.winVersion.major, winVersion.winVersion.minor) == (6, 1):`

I prefer

`if (winVersion.winVersion.major, winVersion.winVersion.minor) < (6, 2):`

**Comment by josephsl** [comment]
Hi, the reason for checking for equality was because we support Windows 7 SP1 and higher. I’ll edit this to make it clearer. Thanks.

**Comment by francipvb** [comment]
What about Version fields such major, minor and revision?

**Comment by josephsl** [comment]
Hi, not quite possible due to so many versioning schemes in use (for example, major.minor, year.major, major.minor.build, major.minor.revision, etc). Thanks.

**Comment by francipvb** [comment]
Hello @josephsl,

> Hi, not quite possible due to so many versioning schemes in use (for example, major.minor, year.major, major.minor.build, major.minor.revision, etc). Thanks.

Anyway, it can be nice to have such features, regardless the version scheme used by the application. Repeating code to parse the actual version of the running application for every app module that requires it is not a good way to do.

Another thing to consider is that the scheme major.minor is a standard under windows and it is relatively safe to add major and minor.

Cheers,

**Comment by josephsl** [comment]
Hi, technically, it is possible to obtain this information. But imagine a case where a user may wish to find out NVDA’s own version via app module methods. Although it is possible to use versionInfo module for this, at first glance, due to the way version string is constructed, major.minor isn’t obvious. Also, note that what is considered major.minor could be year.major i.e. for NVDA, appModule.major will be reported as 2019 instead of 2, for instance. Not that I’m against this idea – I want to see various cases discussed before proceeding with this (by the way, I have suspended work on a PR for this due to high priority work at the moment).

**Comment by francipvb** [comment]
Hello,

El 14/6/2019 a las 12:16, Joseph Lee escribió:
> Hi, technically, it is possible to obtain this information. But imagine
> a case where a user may wish to find out NVDA’s own version via app
> module methods. Although it is possible to use versionInfo module for
> this, at first glance, due to the way version string is constructed,
> major.minor isn’t obvious. Also, note that what is considered
> major.minor could be year.major i.e. for NVDA, appModule.major will be
> reported as 2019 instead of 2, for instance. Not that I’m against this
> idea – I want to see various cases discussed before proceeding with this
> (by the way, I have suspended work on a PR for this due to high priority
> work at the moment).

What about naming these "first", "mid" and "last" or something like? I
think that these three parts are the most widely queried.

Probably the best approach is to implement a new lightweight class to
store these values, constructed from a plain string such productVersion
ones, replacing it by this hypothetical structure.

Sorry for the private responses, I was missing to respond to Github
notification email.

Cheers,
-- 
Cuando tus fuerzas terminan, las de mi Dios comienzan.

File: source/appModuleHandler.py [hunk 2]
```diff
@@ -404,6 +404,16 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
 	# optimisation: Make it easy to detect that this hasn't been overridden.
 	chooseNVDAObjectOverlayClasses._isBase = True
 
+	def _get_appPath(self):
+		"""Returns the full path for the executable e.g. 'C:\\Windows\\explorer.exe' for Explorer.
```
Comment (francipvb) [inline source/appModuleHandler.py]: Sorry but I don't see the utility of this under an appModule.

**Comment by josephsl** [comment]
Hi,

Version class: this can be fetched easily for immersive apps as all that's required will be parsing the user mode app model structure.

Regarding first, mid, last: there are three issues with this:

1. Not exactly user friendly: although great for developers who knows what they stand for, for users, they may wish to write it as appModule.version_major instead of appModule.version_first as major/minor/build/revision terms are more prominent.
2. Does not describe schemes fully: what if the only version info expressions are first, first.last, first.mid, mid.last, etc.? At least version_first is guaranteed, but even then, there are system components that won't return a meaningful version info at all (see below).
3. No version info for processes with NULL handle: in order to fetch product info (or for that matter, anything with processes), a process handle (hProcess) must be obtained. There are situations where this fails, the most prominent example being desktop window object held by CSRSS (Client/Server Runtime Subsystem/Windows subsystem process). At least the PID can be obtained, but when trying to obtain version info, NVDA (or rather, Python) records a traceback, saying that process handle is 0 (note that process handles are relative to the calling application, not system-wide), meaning any proposal regarding parsing version info will fail for CSRSS. There are other system components that won't even show up via Windows API (although they are running), nor if they do show up somehow, hProcess will still be NULL (system process (part of ntoskrnl.exe, PID 0) will refuse to yield version info (for these, you must use sys.getwindowsversion and build versioning info). Note that we cannot catalog all sorts of system processes where hProcess will be NULL - things are changing fast.

Thanks.

**Comment by josephsl** [comment]
Hi,

Actually, yes there is a utility for appPath: virtual machine hosts, executable path for immersive apps, and eventually, for running unit tests.

Thanks.

**Comment by josephsl** [comment]
Hi,

There is a related issue that was discovered while working on Windows 10 Calculator app module: no product info for certain universal apps, and I'm testing a potential solution. Because that issue is different from this PR, an issue/PR pair will be created that will reference this pull request.

Thanks.

**Comment by josephsl** [comment]
Hi,

Update on Windows Store app property: originally it asked user32::IsImmersiveProcess and answered "yes" or "no" depending on return value (1 or 0, respectively). However, this method isn't complete because:

1. What Windows reports as immersive app is actually not the case for File Explorer and other apps. In case of File Explorer, this is because it is a .Net (managed) application.
2. What Windows reports as a native (non-immersive) app turns out to be a Store app, the most notable being Office 365 package obtained from Microsoft Store.

A more accurate way is probing for app package info structure that ships with a Store package. Based on an app model XML, this structure records the following information:

* Package name (publisher.name)
* Product version (a four-part string separated by a dot).
* Architecture: the intended architecture for this app (which can be "neutral" (can run anywhere) or a processor architecture).
* Package language: the language the package is designed for (for many apps, it is either not defined (an empty string) or "neutral").
* App ID: a unique app model identifier for this package.

This structure is returned by kernel32::GetPackageFullName function as a string representation (see #10108 for more details). Every Store package (native or managed, including Store version of NVDA) will return this structure despite immersive process function returning values that do not fully reflect what the underlying package is. There might be a use case for IsImmersiveProcess - perhaps looking to see if we're dealing with a managed application.

Note that the above remark is irrelevant on Windows 7 as it does not come with package info functions nor immersive process checker routine.

Thanks.

**Comment by josephsl** [comment]
Hi,

Suspended until further notice in light of #11006. 

Thanks.

**Comment by Adriani90** [comment]
@josephsl I understand #11006 addresses new pull requests, not already existing pull requests. If we don't continue work on existing pull requests, then we run into the same issue later on.
I suggest if possible, continue work on existing pull requests, so that the backlog can be reviewed and finalized as soon as possible.

**Comment by josephsl** [comment]
Hi, the reason for suspending it for now is to let higher priority PR’s to get a chance to get reviewed. Although this PR is important, it isn’t a high priority work. Thanks.

**Comment by josephsl** [comment]
Hi,

The PR is trivial enough for a fast forward review, but I imagine there might be scenarios where testing might be required. Even then, I think there are PR's that are more urgent than this one (depends on PR tags).

Thanks.

**Comment by feerrenrut** [comment]
> Suspended until further notice in light of #11006.

Please continue to work on this PR. #11006 requests no new PR's. This PR already exists and is thus in the backlog of PR's that we want to address before accepting new ones. We are trying to stop PR's from going stale, and then becoming hard to merge. 

Forgive me for replying to you multiple times, I want to avoid the community being mislead.

**Comment by josephsl** [comment]
Hi, rebasing was successful, so it is ready for review again. Thanks.

File: source/appModuleHandler.py [hunk 3]
```diff
@@ -362,6 +363,22 @@ def _getExecutableFileInfo(self):
...
 
+	def _getImmersivePackageInfo(self):
+		# Used to obtain full package structure for a hosted app.
+		# The package structure consists of product name, version, architecture, language, and app ID.
+		# This is useful for confirming whether an app is hosted or not despite an app reporting otherwise.
+		# Some apps such as File Explorer says it is an immersive process but error 15700 is shown.
+		# Others such as Store version of Office are not truly hosted apps but are distributed via Store.
+		length = ctypes.c_uint()
+		buf = ctypes.windll.kernel32.GetPackageFullName(self.processHandle, ctypes.byref(length), None)
+		packageFullName = ctypes.create_unicode_buffer(buf)
```
Comment (michaelDCurran) [inline source/appModuleHandler.py:374]: The argument to create_unicode_buffer should be length, not buf I believe? 
According to https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getpackagefullname
GetPackageFullname returns ERROR_SUCCESS or an error code. The needed length is written to the length parameter.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

**Comment by josephsl** [comment]
Hi, you’re right. Thanks for clarifying this.

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**